doc/recover.6
.TH RECOVER 6 "14 December 2015" NETHACK
.de ND
.ds Nd \\$3
..
.de NB
.ds Nb \\$2
..
.de NR
.ds Nr \\$2
..
.ND $NHDT-Date: 1524689550 2018/04/25 20:52:30 $
.NB $NHDT-Branch: NetHack-3.6.0 $
.NR $NHDT-Revision: 1.9 $
.ds Na Kenneth Lorber
.SH NAME
recover \- recover a NetHack game interrupted by disaster
.SH SYNOPSIS
.B recover
[
.B \-d
.I directory
]
.I "base1 base2" ...
.SH DESCRIPTION
.PP
Occasionally, a NetHack game will be interrupted by disaster
when the game or the system crashes.
Prior to NetHack v3.1, these games were lost because various information
like the player's inventory was kept only in memory.
Now, all pertinent information can be written out to disk,
so such games can be recovered at the point of the last level change.
.PP
The
.I base
options tell
.I recover
which files to process.
Each base option specifies recovery of a separate game.
.PP
The
.B \-d
option, which must be the first argument if it appears,
supplies a directory which is the NetHack playground.
It overrides the value from NETHACKDIR, HACKDIR, or the directory
specified by the game administrator during compilation
(usually /usr/games/lib/nethackdir).
.PP
^?ALLDOCS
For recovery to be possible,
.I nethack
must have been compiled with the INSURANCE option, and the run-time option
.I checkpoint
must also have been on.
^:
^?INSURANCE
For recovery to be possible,
.I nethack
must have been compiled with the INSURANCE option (this configuration was),
and the run-time option
.I checkpoint
must also have been on.
^:
This configuration of
.I nethack
was created without support for recovery.
^.
^.
NetHack normally writes out files for levels as the player leaves them,
so they will be ready for return visits.
When checkpointing, NetHack also writes out the level entered and
the current game state on every level change.
This naturally slows level changes down somewhat.
.PP
The level file names are of the form base.nn, where nn is an internal
bookkeeping number for the level.
The file base.0 is used for game identity, locking, and, when checkpointing,
for the game state.
Various OSes use different strategies for constructing the base name.
Microcomputers use the character name, possibly truncated and modified
to be a legal filename on that system.
Multi-user systems use the (modified) character name prefixed
by a user number to avoid conflicts,
or "xlock" if the number of concurrent players is being limited.
It may be necessary to look in the playground to find the correct
base name of the interrupted game.
.I recover
will transform these level files into a save file of the same name as
.I nethack
would have used.
.PP
Since
.I recover
must be able to read and delete files from the playground
and create files in the save directory,
it has interesting interactions with game security.
Giving ordinary players access to
.I recover
through setuid or setgid is tantamount to leaving the playground
world-writable,
with respect to both cheating and messing up other players.
For a single-user system, this of course does not change anything,
so some of the microcomputer ports install
.I recover
by default.
.PP
For a multi-user system,
the game administrator may want to arrange for all .0 files in the
playground to be fed to recover when the host machine boots,
and handle game crashes individually.
If the user population is sufficiently trustworthy,
.I recover
can be installed with the same permissions the
.I nethack
executable has.
In either case,
.I recover
is easily compiled from the distribution utility directory.
.SH NOTES
.PP
Like
.I nethack
itself,
.I recover
will overwrite existing savefiles of the same name.
Savefiles created by
.I recover
are uncompressed;
they may be compressed afterwards if desired,
but even a compression-using
.I nethack
will find them in the uncompressed form.
.SH "SEE ALSO"
evilhack(6)
.SH BUGS
.PP
.I recover
makes no attempt to find out if a base name specifies a game in progress.
If multiple machines share a playground, this would be impossible to
determine.
.PP
.I recover
should be taught to use the nethack playground locking mechanism to
avoid conflicts.
.SH COPYRIGHT
This file is Copyright (C) \*(Na and was last modified \*(Nd (version
\*(Nb:\*(Nr).
NetHack may be freely redistributed.  See license for details.


dat/history
EvilHack History file

EvilHack is a variant initially based on NetHack 3.6.2, created by Keith
Simpson.  Work on EvilHack began on October 20th 2018, and was released for
public play in April of 2019 on the Hardfought public NetHack servers.
Official releases, including compiled windows binaries, can be found at the
Github repository for EvilHack - https://github.com/k21971/EvilHack

NetHack's history can be viewed below.

-----------------------------------------------------------------------------

NetHack History file for release 3.6

Behold, mortal, the origins of NetHack...

Jay Fenlason wrote the original Hack with help from Kenny Woodland,
Mike Thome, and Jon Payne.

Andries Brouwer did a major re-write, transforming Hack into a very different
game, and published (at least) three versions (1.0.1, 1.0.2, and 1.0.3) for
UNIX(tm) machines to the Usenet.

Don G. Kneller ported Hack 1.0.3 to Microsoft(tm) C and MS-DOS(tm), producing
PC HACK 1.01e, added support for DEC Rainbow graphics in version 1.03g, and
went on to produce at least four more versions (3.0, 3.2, 3.51, and 3.6;
note that these are old Hack version numbers, not contemporary NetHack ones).

R. Black ported PC HACK 3.51 to Lattice(tm) C and the Atari 520/1040ST,
producing ST Hack 1.03.

Mike Stephenson merged these various versions back together, incorporating
many of the added features, and produced NetHack version 1.4 in 1987.  He
then coordinated a cast of thousands in enhancing and debugging NetHack 1.4
and released NetHack versions 2.2 and 2.3.

Later, Mike coordinated a major rewrite of the game, heading a team which
included Ken Arromdee, Jean-Christophe Collet, Steve Creps, Eric Hendrickson,
Izchak Miller, Eric S. Raymond, John Rupley, Mike Threepoint, and Janet Walz,
to produce NetHack 3.0c.  The same group subsequently released ten patch-
level revisions and updates of 3.0.

NetHack 3.0 was ported to the Atari by Eric R. Smith, to OS/2 by Timo
Hakulinen, and to VMS by David Gentzel.  The three of them and Kevin Darcy
later joined the main NetHack Development Team to produce subsequent
revisions of 3.0.

Olaf Seibert ported NetHack 2.3 and 3.0 to the Amiga.  Norm Meluch, Stephen
Spackman and Pierre Martineau designed overlay code for PC NetHack 3.0.
Johnny Lee ported NetHack 3.0 to the Macintosh.  Along with various other
Dungeoneers, they continued to enhance the PC, Macintosh, and Amiga ports
through the later revisions of 3.0.

Headed by Mike Stephenson and coordinated by Izchak Miller and Janet Walz,
the NetHack Development Team which now included Ken Arromdee, David Cohrs,
Jean-Christophe Collet, Kevin Darcy, Matt Day, Timo Hakulinen, Steve Linhart,
Dean Luick, Pat Rankin, Eric Raymond, and Eric Smith undertook a radical
revision of 3.0.  They re-structured the game's design, and re-wrote major
parts of the code.  They added multiple dungeons, a new display, special
individual character quests, a new endgame and many other new features, and
produced NetHack 3.1.

Ken Lorber, Gregg Wonderly and Greg Olson, with help from Richard Addison,
Mike Passaretti, and Olaf Seibert, developed NetHack 3.1 for the Amiga.

Norm Meluch and Kevin Smolkowski, with help from Carl Schelin, Stephen
Spackman, Steve VanDevender, and Paul Winner, ported NetHack 3.1 to the PC.

Jon W{tte and Hao-yang Wang, with help from Ross Brown, Mike Engber, David
Hairston, Michael Hamel, Jonathan Handler, Johnny Lee, Tim Lennan, Rob Menke,
and Andy Swanson developed NetHack 3.1 for the Macintosh, porting it for
MPW.  Building on their development, Bart House added a Think C port.

Timo Hakulinen ported NetHack 3.1 to OS/2.  Eric Smith ported NetHack 3.1
to the Atari.  Pat Rankin, with help from Joshua Delahunty, is responsible
for the VMS version of NetHack 3.1.  Michael Allison ported NetHack 3.1 to
Windows NT.

Dean Luick, with help from David Cohrs, developed NetHack 3.1 for X11.
Warwick Allison wrote a tiled version of NetHack for the Atari;
he later contributed the tiles to the NetHack Development Team and tile
support was then added to other platforms.

The 3.2 NetHack Development Team, comprised of Michael Allison, Ken Arromdee, 
David Cohrs, Jessie Collet, Steve Creps, Kevin Darcy, Timo Hakulinen, Steve
Linhart, Dean Luick, Pat Rankin, Eric Smith, Mike Stephenson, Janet Walz,
and Paul Winner, released version 3.2 in April of 1996.

Version 3.2 marked the tenth anniversary of the formation of the development
team.  In a testament to their dedication to the game, all thirteen members
of the original NetHack Development Team remained on the team at the start of
work on that release.  During the interval between the release of 3.1.3 and
3.2, one of the founding members of the NetHack Development Team, 
Dr. Izchak Miller, passed away.  That release of the game was dedicated to
him by the development and porting teams.

Version 3.2 proved to be more stable than previous versions.  Many bugs
were fixed, abuses eliminated, and game features tuned for better game
play.

During the lifespan of NetHack 3.1 and 3.2, several enthusiasts of the game
added their own modifications to the game and made these "variants" publicly
available:

Tom Proudfoot and Yuval Oren created NetHack++, which was quickly renamed
NetHack--.  Working independently, Stephen White wrote NetHack Plus.
Tom Proudfoot later merged NetHack Plus and his own NetHack-- to produce
SLASH.  Larry Stewart-Zerba and Warwick Allison improved the spellcasting
system with the Wizard Patch.  Warwick Allison also ported NetHack to use
the Qt interface.

Warren Cheung combined SLASH with the Wizard Patch to produce Slash'em, and
with the help of Kevin Hugo, added more features.  Kevin later joined the
NetHack Development Team and incorporated the best of these ideas in 
NetHack 3.3.

The final update to 3.2 was the bug fix release 3.2.3, which was released
simultaneously with 3.3.0 in December 1999 just in time for the Year 2000.


util/recover.c
/* NetHack 3.6	recover.c	$NHDT-Date: 1550103078 2019/02/14 00:11:18 $  $NHDT-Branch: NetHack-3.6.2-beta01 $:$NHDT-Revision: 1.19 $ */
/*	Copyright (c) Janet Walz, 1992.				  */
/* NetHack may be freely redistributed.  See license for details. */

/*
 *  Utility for reconstructing NetHack save file from a set of individual
 *  level files.  Requires that the `checkpoint' option be enabled at the
 *  time NetHack creates those level files.
 */
#ifdef WIN32
#include <errno.h>
#include "win32api.h"
#endif

#include "config.h"
#if !defined(O_WRONLY) && !defined(LSC) && !defined(AZTEC_C)
#include <fcntl.h>
#endif

#ifdef VMS
extern int FDECL(vms_creat, (const char *, unsigned));
extern int FDECL(vms_open, (const char *, int, unsigned));
#endif /* VMS */

int FDECL(restore_savefile, (char *, const char *));
static void FDECL(set_levelfile_name, (int));
static int FDECL(open_levelfile, (int, const char *));
static int FDECL(create_savefile, (const char *));
void FDECL(copy_bytes, (int, int));

#ifndef WIN_CE
#define Fprintf (void) fprintf
#else
#define Fprintf (void) nhce_message
static void nhce_message(FILE *, const char *, ...);
#endif

#define Close (void) close

#ifdef UNIX
#define SAVESIZE (PL_NSIZ + 13) /* save/99999player.e */
#else
#ifdef VMS
#define SAVESIZE (PL_NSIZ + 22) /* [.save]<uid>player.e;1 */
#else
#ifdef WIN32
#define SAVESIZE (PL_NSIZ + 40) /* username-player.NetHack-saved-game */
#else
#define SAVESIZE FILENAME /* from macconf.h or pcconf.h */
#endif
#endif
#endif

#if defined(EXEPATH)
char *FDECL(exepath, (char *));
#endif

#if defined(__BORLANDC__) && !defined(_WIN32)
extern unsigned _stklen = STKSIZ;
#endif
char savename[SAVESIZE]; /* holds relative path of save file from playground */

#ifndef NO_MAIN
int
main(argc, argv)
int argc;
char *argv[];
{
    int argno;
    const char *dir = (char *) 0;
#ifdef AMIGA
    char *startdir = (char *) 0;
#endif

    if (!dir)
        dir = getenv("NETHACKDIR");
    if (!dir)
        dir = getenv("HACKDIR");
#if defined(EXEPATH)
    if (!dir)
        dir = exepath(argv[0]);
#endif
    if (argc == 1 || (argc == 2 && !strcmp(argv[1], "-"))) {
        Fprintf(stderr, "Usage: %s [ -d directory ] base1 [ base2 ... ]\n",
                argv[0]);
#if defined(WIN32) || defined(MSDOS)
        if (dir) {
            Fprintf(
                stderr,
                "\t(Unless you override it with -d, recover will look \n");
            Fprintf(stderr, "\t in the %s directory on your system)\n", dir);
        }
#endif
        exit(EXIT_FAILURE);
    }

    argno = 1;
    if (!strncmp(argv[argno], "-d", 2)) {
        dir = argv[argno] + 2;
        if (*dir == '=' || *dir == ':')
            dir++;
        if (!*dir && argc > argno) {
            argno++;
            dir = argv[argno];
        }
        if (!*dir) {
            Fprintf(stderr,
                    "%s: flag -d must be followed by a directory name.\n",
                    argv[0]);
            exit(EXIT_FAILURE);
        }
        argno++;
    }
#if defined(SECURE) && !defined(VMS)
    if (dir
#ifdef HACKDIR
        && strcmp(dir, HACKDIR)
#endif
            ) {
        (void) setgid(getgid());
        (void) setuid(getuid());
    }
#endif /* SECURE && !VMS */

#ifdef HACKDIR
    if (!dir)
        dir = HACKDIR;
#endif

#ifdef AMIGA
    startdir = getcwd(0, 255);
#endif
    if (dir && chdir((char *) dir) < 0) {
        Fprintf(stderr, "%s: cannot chdir to %s.\n", argv[0], dir);
        exit(EXIT_FAILURE);
    }

    while (argc > argno) {
        if (restore_savefile(argv[argno], dir) == 0)
            Fprintf(stderr, "recovered \"%s\" to %s\n", argv[argno],
                    savename);
        argno++;
    }
#ifdef AMIGA
    if (startdir)
        (void) chdir(startdir);
#endif
    exit(EXIT_SUCCESS);
    /*NOTREACHED*/
    return 0;
}
#endif /* !NO_MAIN */

static char lock[256];

static void
set_levelfile_name(lev)
int lev;
{
    char *tf;

    tf = rindex(lock, '.');
    if (!tf)
        tf = lock + strlen(lock);
    (void) sprintf(tf, ".%d", lev);
#ifdef VMS
    (void) strcat(tf, ";1");
#endif
}

static int
open_levelfile(lev, directory)
int lev;
const char *directory;
{
    int fd;
    char levelfile[BUFSIZ];

    set_levelfile_name(lev);
    if (directory) {
        snprintf(levelfile, BUFSIZ, "%s/%s", directory, lock);
    } else {
        strcpy(levelfile, lock);
    }
#if defined(MICRO) || defined(WIN32) || defined(MSDOS)
    fd = open(levelfile, O_RDONLY | O_BINARY);
#else
    fd = open(levelfile, O_RDONLY, 0);
#endif
    return fd;
}

static int
create_savefile(directory)
const char *directory;
{
    int fd;
    char savefile[BUFSIZ];

    if (directory) {
        sprintf(savefile, "%s/%s", directory, savename);
    } else {
        strcpy(savefile, savename);
    }
#if defined(MICRO) || defined(WIN32) || defined(MSDOS)
    fd = open(savefile, O_WRONLY | O_BINARY | O_CREAT | O_TRUNC, FCMASK);
#else
    fd = creat(savefile, FCMASK);
#endif
    return fd;
}

void
copy_bytes(ifd, ofd)
int ifd, ofd;
{
    char buf[BUFSIZ];
    int nfrom, nto;

    do {
        nfrom = read(ifd, buf, BUFSIZ);
        nto = write(ofd, buf, nfrom);
        if (nto != nfrom) {
            Fprintf(stderr, "file copy failed!\n");
            exit(EXIT_FAILURE);
        }
    } while (nfrom == BUFSIZ);
}

int
restore_savefile(basename, directory)
char *basename;
const char *directory;
{
    int gfd, lfd, sfd;
    int res = 0, lev, savelev, hpid, pltmpsiz;
    xchar levc;
    struct version_info version_data;
    struct savefile_info sfi;
    char plbuf[PL_NSIZ];

    /* level 0 file contains:
     *  pid of creating process (ignored here)
     *  level number for current level of save file
     *  name of save file nethack would have created
     *  savefile info
     *  player name
     *  and game state
     */
    (void) strcpy(lock, basename);
    gfd = open_levelfile(0, directory);
    if (gfd < 0) {
#if defined(WIN32) && !defined(WIN_CE)
        if (errno == EACCES) {
            Fprintf(
                stderr,
                "\nThere are files from a game in progress under your name.");
            Fprintf(stderr, "\nThe files are locked or inaccessible.");
            Fprintf(stderr, "\nPerhaps the other game is still running?\n");
        } else
            Fprintf(stderr, "\nTrouble accessing level 0 (errno = %d).\n",
                    errno);
#endif
        Fprintf(stderr, "Cannot open level 0 for %s.\n", basename);
        return -1;
    }
    if (read(gfd, (genericptr_t) &hpid, sizeof hpid) != sizeof hpid) {
        Fprintf(
            stderr, "%s\n%s%s%s\n",
            "Checkpoint data incompletely written or subsequently clobbered;",
            "recovery for \"", basename, "\" impossible.");
        Close(gfd);
        return -1;
    }
    if (read(gfd, (genericptr_t) &savelev, sizeof(savelev))
        != sizeof(savelev)) {
        Fprintf(stderr, "Checkpointing was not in effect for %s -- recovery "
                        "impossible.\n",
                basename);
        Close(gfd);
        return -1;
    }
    if ((read(gfd, (genericptr_t) savename, sizeof savename)
         != sizeof savename)
        || (read(gfd, (genericptr_t) &version_data, sizeof version_data)
            != sizeof version_data)
        || (read(gfd, (genericptr_t) &sfi, sizeof sfi) != sizeof sfi)
        || (read(gfd, (genericptr_t) &pltmpsiz, sizeof pltmpsiz)
            != sizeof pltmpsiz) || (pltmpsiz > PL_NSIZ)
        || (read(gfd, (genericptr_t) plbuf, pltmpsiz) != pltmpsiz)) {
        Fprintf(stderr, "Error reading %s -- can't recover.\n", lock);
        Close(gfd);
        return -1;
    }

    /* save file should contain:
     *  version info
     *  savefile info
     *  player name
     *  current level (including pets)
     *  (non-level-based) game state
     *  other levels
     */
    sfd = create_savefile(directory);
    if (sfd < 0) {
        Fprintf(stderr, "Cannot create savefile %s.\n", savename);
        Close(gfd);
        return -1;
    }

    lfd = open_levelfile(savelev, directory);
    if (lfd < 0) {
        Fprintf(stderr, "Cannot open level of save for %s.\n", basename);
        Close(gfd);
        Close(sfd);
        return -1;
    }

    if (write(sfd, (genericptr_t) &version_data, sizeof version_data)
        != sizeof version_data) {
        Fprintf(stderr, "Error writing %s; recovery failed.\n", savename);
        Close(gfd);
        Close(sfd);
        Close(lfd);
        return -1;
    }

    if (write(sfd, (genericptr_t) &sfi, sizeof sfi) != sizeof sfi) {
        Fprintf(stderr,
                "Error writing %s; recovery failed (savefile_info).\n",
                savename);
        Close(gfd);
        Close(sfd);
        Close(lfd);
        return -1;
    }

    if (write(sfd, (genericptr_t) &pltmpsiz, sizeof pltmpsiz)
        != sizeof pltmpsiz) {
        Fprintf(stderr,
                "Error writing %s; recovery failed (player name size).\n",
                savename);
        Close(gfd);
        Close(sfd);
        Close(lfd);
        return -1;
    }

    if (write(sfd, (genericptr_t) plbuf, pltmpsiz) != pltmpsiz) {
        Fprintf(stderr, "Error writing %s; recovery failed (player name).\n",
                savename);
        Close(gfd);
        Close(sfd);
        Close(lfd);
        return -1;
    }

    copy_bytes(lfd, sfd);
    Close(lfd);
    (void) unlink(lock);

    copy_bytes(gfd, sfd);
    Close(gfd);
    set_levelfile_name(0);
    (void) unlink(lock);

    for (lev = 1; lev < 256 && res == 0; lev++) {
        /* level numbers are kept in xchars in save.c, so the
         * maximum level number (for the endlevel) must be < 256
         */
        if (lev != savelev) {
            lfd = open_levelfile(lev, directory);
            if (lfd >= 0) {
                /* any or all of these may not exist */
                levc = (xchar) lev;
                if (write(sfd, (genericptr_t) &levc, sizeof levc)
                    != sizeof levc)
                    res = -1;
                else
                    copy_bytes(lfd, sfd);
                Close(lfd);
                (void) unlink(lock);
            }
        }
    }

    Close(sfd);

#if 0 /* OBSOLETE, HackWB is no longer in use */
#ifdef AMIGA
    if (res == 0) {
        /* we need to create an icon for the saved game
         * or HackWB won't notice the file.
         */
        char iconfile[FILENAME];
        int in, out;

        (void) sprintf(iconfile, "%s.info", savename);
        in = open("NetHack:default.icon", O_RDONLY);
        out = open(iconfile, O_WRONLY | O_TRUNC | O_CREAT);
        if (in > -1 && out > -1) {
            copy_bytes(in, out);
        }
        if (in > -1)
            close(in);
        if (out > -1)
            close(out);
    }
#endif /*AMIGA*/
#endif
    return res;
}

#ifdef EXEPATH
#ifdef __DJGPP__
#define PATH_SEPARATOR '/'
#else
#define PATH_SEPARATOR '\\'
#endif

#define EXEPATHBUFSZ 256
char exepathbuf[EXEPATHBUFSZ];

char *
exepath(str)
char *str;
{
    char *tmp, *tmp2;
    int bsize;

    if (!str)
        return (char *) 0;
    bsize = EXEPATHBUFSZ;
    tmp = exepathbuf;
#if !defined(WIN32)
    strcpy(tmp, str);
#else
#if defined(WIN_CE)
    {
        TCHAR wbuf[EXEPATHBUFSZ];
        GetModuleFileName((HANDLE) 0, wbuf, EXEPATHBUFSZ);
        NH_W2A(wbuf, tmp, bsize);
    }
#else
    *(tmp + GetModuleFileName((HANDLE) 0, tmp, bsize)) = '\0';
#endif
#endif
    tmp2 = strrchr(tmp, PATH_SEPARATOR);
    if (tmp2)
        *tmp2 = '\0';
    return tmp;
}
#endif /* EXEPATH */

#ifdef AMIGA
#include "date.h"
const char amiga_version_string[] = AMIGA_VERSION_STRING;
#endif

#ifdef WIN_CE
void
nhce_message(FILE *f, const char *str, ...)
{
    va_list ap;
    TCHAR wbuf[NHSTR_BUFSIZE];
    char buf[NHSTR_BUFSIZE];

    va_start(ap, str);
    vsprintf(buf, str, ap);
    va_end(ap);

    MessageBox(NULL, NH_A2W(buf, wbuf, NHSTR_BUFSIZE), TEXT("Recover"),
               MB_OK);
}
#endif

/*recover.c*/
